### PR TITLE
Invariant tests

### PR DESCRIPTION
**Pull request by dylandesrosier**
Also fixes several issues with events, sponsorship and safety checks for periods.

### CODE REVIEW

File: foundry.toml [hunk 1]
```diff
@@ -6,6 +6,12 @@ libs = ['lib']
 solc = "0.8.17"
 fs_permissions = [{ access = "read", path = "./broadcast"}]
 gas_reports = ["*"]
+no_match_test = "SLOW"
+
+[invariant]
+fail_on_revert = false
+# depth = 32
+# runs = 1
```
Comment (PierrickGT) [inline foundry.toml]: Remove these params if they're not being used.

File: src/TwabController.sol [hunk 2]
```diff
@@ -423,48 +463,64 @@ contract TwabController {
...
       _decreaseBalances(_vault, _from, _amount, _isFromDelegate ? _amount : 0);
 
-      if (!_isFromDelegate) {
-        _decreaseBalances(
-          _vault,
-          _fromDelegate,
-          0,
-          _fromDelegate != SPONSORSHIP_ADDRESS ? _amount : 0
-        );
+      // If the user is not delegating to themself, increase the delegate's delegateBalance
```
Comment (PierrickGT) [inline src/TwabController.sol]: It should be `decrease` instead of `increase` no?
```suggestion
      // If the user is not delegating to themself, decrease the delegate's delegateBalance
```

File: src/TwabController.sol [hunk 3]
```diff
@@ -423,48 +463,64 @@ contract TwabController {
...
+      // If the user is delegating to the sponsorship address, don't adjust the delegateBalance
```
Comment (PierrickGT) [inline src/TwabController.sol:474]: This condition is already handled above no?
If `_isFromDelegate` is `true`, we decrease the `delegateBalance`, otherwise we don't.
That being said here, you handle the `delegateBalance` of the `fromDelegate` but this is handled in `_transferDelegateBalance`.

File: src/TwabController.sol [hunk 4]
```diff
@@ -423,48 +463,64 @@ contract TwabController {
...
+      if (!_isFromDelegate && _fromDelegate != SPONSORSHIP_ADDRESS) {
+        _decreaseBalances(_vault, _fromDelegate, 0, _amount);
```
Comment (PierrickGT) [inline src/TwabController.sol:476]: It's exactly what we do in the `_transferDelegateBalance` function:
https://github.com/pooltogether/v5-twab-controller/blob/70e9ada002dd23ed6af27b062588bdcf91266105/src/TwabController.sol#L565

So maybe this function should be called in `mint`, `burn` and `transfer` if `_isFromDelegate` is `false`.

File: src/TwabController.sol [hunk 5]
```diff
@@ -423,48 +463,64 @@ contract TwabController {
...
+        (_toDelegate == SPONSORSHIP_ADDRESS && _fromDelegate != SPONSORSHIP_ADDRESS)
+      ) {
+        // If the user is delegating to the sponsorship address, don't adjust the total supply delegateBalance
         _decreaseTotalSupplyBalances(
           _vault,
-          _amount,
-          _fromDelegate != SPONSORSHIP_ADDRESS ? _amount : 0
+          _to == address(0) ? _amount : 0,
+          (_to == address(0) && _fromDelegate != SPONSORSHIP_ADDRESS) ||
+            (_toDelegate == SPONSORSHIP_ADDRESS && _fromDelegate != SPONSORSHIP_ADDRESS)
```
Comment (PierrickGT) [inline src/TwabController.sol:491]: Why do we need `_toDelegate` here? 
Transfers should only be `from` -> `to`, so we shouldn't have to care about the `toDelegate`.
Plus, when we delegate to the `SPONSORSHIP_ADDRESS`, the `delegateBalance` of the `totalSupply` should not change since we are supposed to have the same amount of token in the system.

File: src/TwabController.sol [hunk 6]
```diff
@@ -423,48 +463,64 @@ contract TwabController {
...
       bool _isToDelegate = _toDelegate == _to;
 
+      // If the user is delegating to themself, increase their delegateBalance
       _increaseBalances(_vault, _to, _amount, _isToDelegate ? _amount : 0);
 
-      if (!_isToDelegate) {
-        _increaseBalances(_vault, _toDelegate, 0, _toDelegate != SPONSORSHIP_ADDRESS ? _amount : 0);
+      // Otherwise, increase their delegates delegateBalance if it is not the sponsorship address
+      if (!_isToDelegate && _toDelegate != SPONSORSHIP_ADDRESS) {
+        _increaseBalances(_vault, _toDelegate, 0, _amount);
```
Comment (PierrickGT) [inline src/TwabController.sol:507]: Same remark, I feel like it should be handled by the `_transferDelegateBalance` function.

File: src/TwabController.sol [hunk 7]
```diff
@@ -423,48 +463,64 @@ contract TwabController {
...
+        _from == address(0) ||
+        (_fromDelegate == SPONSORSHIP_ADDRESS && _toDelegate != SPONSORSHIP_ADDRESS)
+      ) {
         _increaseTotalSupplyBalances(
           _vault,
-          _amount,
-          _toDelegate != SPONSORSHIP_ADDRESS ? _amount : 0
+          _from == address(0) ? _amount : 0,
+          (_from == address(0) && _toDelegate != SPONSORSHIP_ADDRESS) ||
+            (_fromDelegate == SPONSORSHIP_ADDRESS && _toDelegate != SPONSORSHIP_ADDRESS)
```
Comment (PierrickGT) [inline src/TwabController.sol:521]: Same remark regarding the `SPONSORSHIP_ADDRESS`, the `delegateBalance` of the `totalSupply` should not change.

File: src/libraries/TwabLib.sol [hunk 8]
```diff
@@ -639,8 +645,13 @@ library TwabLib {
       return false;
     }
     // If there is one observation, compare it's timestamp
+    uint32 period = _getTimestampPeriod(_time);
```
Comment (PierrickGT) [inline src/libraries/TwabLib.sol:648]: You only use `period` once no? Not sure you need to assign it to a variable.

File: test/TwabController.t.sol [hunk 9]
```diff
@@ -876,9 +923,90 @@ contract TwabControllerTest is BaseSetup {
...
+    );
+    totalSupplyTwab = twabController.getTotalSupplyTwabBetween(
+      mockVault,
+      TwabLib.PERIOD_OFFSET,
+      newestObservation.timestamp
+    );
+
+    assertEq(aliceTwab, bobTwab);
+    // Sum of Alice and Bob's TWABs should be less than or equal to the total supply TWAB
+    assertLe(aliceTwab + bobTwab, totalSupplyTwab);
```
Comment (PierrickGT) [inline test/TwabController.t.sol]: Shouldn't it be exactly equal to the total supply TWAB?

File: test/TwabController.t.sol [hunk 10]
```diff
@@ -876,9 +923,90 @@ contract TwabControllerTest is BaseSetup {
...
+
+    (, newestObservation) = twabController.getNewestTotalSupplyObservation(mockVault);
+    (, newestAliceObservation) = twabController.getNewestObservation(mockVault, alice);
+    (, newestBobObservation) = twabController.getNewestObservation(mockVault, bob);
+
+    if (newestObservation.timestamp < newestAliceObservation.timestamp) {
+      newestObservation = newestAliceObservation;
+    }
+    if (newestObservation.timestamp < newestBobObservation.timestamp) {
+      newestObservation = newestBobObservation;
```
Comment (PierrickGT) [inline test/TwabController.t.sol:1006]: `newestObservation` will always be equal to `newestBobObservation` since Bob observation is always updated last?

File: test/TwabLib.t.sol [hunk 11]
```diff
@@ -683,9 +714,20 @@ contract TwabLibTest is BaseSetup {
...
     uint32 t3 = TwabLib.PERIOD_OFFSET + (TwabLib.PERIOD_LENGTH * 3);
+    ObservationLib.Observation memory prevOrAtObservation;
 
     vm.warp(t0);
     twabLibMock.increaseBalances(1, 1);
+
+    // Get observation at timestamp before first observation
+    prevOrAtObservation = twabLibMock.getPreviousOrAtObservation(t0 - 1 seconds);
+    assertEq(prevOrAtObservation.timestamp, TwabLib.PERIOD_OFFSET);
+    assertGe(prevOrAtObservation.timestamp, TwabLib.PERIOD_OFFSET);
```
Comment (PierrickGT) [inline test/TwabLib.t.sol]: We can only keep `assertEq` here.

```suggestion
    assertEq(prevOrAtObservation.timestamp, TwabLib.PERIOD_OFFSET);
```

File: test/invariants/TwabControllerInvariant.t.sol [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+  // The sum of user balances always equals the total supply.
+  // The sum of user balances always equals total minted - total burned.
+  function invariant_totalSupplyMatchesUserBalances() public {
+    uint256 totalUsersSupply = handler.sumUserBalancesAcrossVaults();
+    uint256 totalTotalSupply = handler.sumTotalSupplyAcrossVaults();
+
+    assertEq(totalUsersSupply, handler.h_totalMinted() - handler.h_totalBurned());
+
+    assertEq(totalUsersSupply, totalTotalSupply);
+    // assertAlmostEqual(totalUsersSupply, totalTotalSupply, 1);
```
Comment (PierrickGT) [inline test/invariants/TwabControllerInvariant.t.sol]: Can be removed.

File: test/invariants/TwabControllerInvariant.t.sol [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+
+  // The sum of User TWABs across the whole time range is less than or equal to the total supply TWAB.
+  function invariant_totalSupplyTwabGreaterThanUserTwabs() public {
+    (uint256 totalSupplyTwabSum, uint256 userTwabSum) = handler.reduceFullRangeTwabs();
+
+    // NOTE: Users odds will be squished and the full odds space won't be allocated.abi
+    assertLe(userTwabSum, totalSupplyTwabSum);
+  }
+
+  /* ============ Helpers ============ */
```
Comment (PierrickGT) [inline test/invariants/TwabControllerInvariant.t.sol]: ✂️

File: test/invariants/handlers/TwabControllerHandler.sol [hunk 14]
```diff
@@ -0,0 +1,427 @@
...
+    uint256 vaultSeed,
+    uint256 actorSeed,
+    address to,
+    uint256 numSeconds
+  ) public useVaultAndActorWithBalance(vaultSeed, actorSeed) countCall("transfer") {
+    timeTravel(numSeconds);
+
+    require(to != address(0) && to != address(1), "Invalid to address");
+
+    amount = uint96(bound(amount, 0, twabController.balanceOf(currentVault, currentActor)));
```
Comment (PierrickGT) [inline test/invariants/handlers/TwabControllerHandler.sol:97]: If the transfer is 0, it should revert no?
At least I think that's how it works in the ERC20 implementation of OZ, not sure we have implemented this safe guard for the TWAB.

File: test/invariants/handlers/TwabControllerHandler.sol [hunk 15]
```diff
@@ -0,0 +1,427 @@
...
+
+    // Manually add the new actor so they are tracked.
+    if (amount > 0) {
```
Comment (PierrickGT) [inline test/invariants/handlers/TwabControllerHandler.sol:100]: Would also avoid this check if the minimum bound value was 1.

File: test/invariants/handlers/TwabControllerHandler.sol [hunk 16]
```diff
@@ -0,0 +1,427 @@
...
+        account = twabController.getTotalSupplyAccount(vault);
+        (, newestObservation) = twabController.getNewestTotalSupplyObservation(vault);
+
+        // Check if the newest observation is safe for this vault
+        isVaultsSafe =
+          isVaultsSafe &&
+          twabController.isTotalSupplyTimeSafe(vault, newestObservation.timestamp);
+
+        // Check if each observation is safe for this vault
+        for (uint256 o_i; o_i < account.details.cardinality; ++o_i) {
```
Comment (PierrickGT) [inline test/invariants/handlers/TwabControllerHandler.sol:278]: Even optimizing for gas in tests?

File: test/utils/BaseTest.sol [hunk 17]
```diff
@@ -25,21 +25,27 @@ contract BaseSetup is Test {
...
+    console.log("-- Observations --");
     for (uint256 i = 0; i < amount; i++) {
       ObservationLib.Observation memory observation = account.observations[i];
-      console.log("Observation: ", i, observation.cumulativeBalance, observation.timestamp);
+      console.log(i, observation.cumulativeBalance, observation.balance, observation.timestamp);
     }
     console.log("--");
   }
+
+  function assertAlmostEqual(uint256 a, uint256 b, uint256 maxDelta) internal {
```
Comment (PierrickGT) [inline test/utils/BaseTest.sol:47]: I don't think this function is needed.
You can use [assertApproxEqAbs](https://book.getfoundry.sh/reference/forge-std/assertApproxEqAbs?highlight=assertApprox#assertapproxeqabs) if you want to test values with a delta.

**Review by PierrickGT — CHANGES_REQUESTED** [review]

File: src/TwabController.sol (hunk 3 above)
Comment (dylandesrosier) [reply src/TwabController.sol:474]: No it isn't caught above. The first `_decreaseBalances` updates the users balances. The second updates their delegates if necessary. So it's removing the delegation from the  `fromDelegate`.

`_transferDelegateBalance` is only used when delegating/sponsoring, it's just the delegation balance portion of `_transfer` so it's a little cheaper.

File: src/libraries/TwabLib.sol (hunk 8 above)
Comment (dylandesrosier) [reply src/libraries/TwabLib.sol:648]: No, it's used in the case of `cardinality == 1` as well as the regular return.

File: test/TwabController.t.sol (hunk 10 above)
Comment (dylandesrosier) [reply test/TwabController.t.sol:1006]: Alice, Bob and the Total Supply are all updated at the same timestamp in the `testTwabEquivalence` test. But the `assertTwabEquivalence` function can be reused (and was previously) so it's nice to have the proper search for the newest.

File: test/invariants/handlers/TwabControllerHandler.sol (hunk 14 above)
Comment (dylandesrosier) [reply test/invariants/handlers/TwabControllerHandler.sol:97]: I don't see any checks for 0 in the OZ ERC20. <https://github.com/OpenZeppelin/openzeppelin-contracts/blob/5420879d9b834a0579423d668fb60c5fc13b60cc/contracts/token/ERC20/ERC20.sol#L222>

It's currently not implemented, we could though

File: test/invariants/handlers/TwabControllerHandler.sol (hunk 16 above)
Comment (dylandesrosier) [reply test/invariants/handlers/TwabControllerHandler.sol:278]: Not quite sure what you mean.
The exact timestamp of each observation recorded should fall within the "safe" time range though, that's what the check inside this `o_i` loop is for.
https://dev.pooltogether.com/protocol/next/design/twab-controller#historic-balances-accuracy

File: test/utils/BaseTest.sol (hunk 17 above)
Comment (dylandesrosier) [reply test/utils/BaseTest.sol:47]: 🤦 I was looking through the wrong section in the docs for the approximate checks.. <https://book.getfoundry.sh/reference/ds-test>

File: src/TwabController.sol (hunk 5 above)
Comment (dylandesrosier) [reply src/TwabController.sol:491]: `delegateBalance` != `balance`. It is the amount of tokens delegated, not the amount of tokens. The amount of tokens in the system stays the same, but the amount delegated can change based on the delegates of the `to` and `from` users.

We don't want to count delegations to the `SPONSORSHIP_ADDRESS`, so the `delegateBalance` of it should always be `0`, meaning delegations to the `SPONSORSHIP_ADDRESS` have no impact on the total supply `delegateBalance`.

So if Alice is delegating 10 to herself and Bob is delegating 10 to `SPONSORSHIP_ADDRESS`, the total supply `delegateBalance` is just 10 (Alice). If Alice then transfers 10 to Bob we need to transfer the `delegateBalance` from Alice to Bob. But since he is delegating to `SPONSORSHIP_ADDRESS` we're really burning the 10 from Alice and need to subtract from the total supply `delegateBalance`.

File: src/TwabController.sol (hunk 6 above)
Comment (dylandesrosier) [reply src/TwabController.sol:507]: `_transferDelegateBalance` is used when we delegate. 

If we called `_transferDelegateBalance` inside `_transferBalance` we would be duplicating all of the checks when we need to do them for the balance transfer. 
If the user is delegating to themself we only need to do one write to memory to update both values. If we always break up the updates to `balance` and `delegateBalance` we would need to do 2 writes if the user is delegating to themself.

File: test/TwabController.t.sol (hunk 9 above)
Comment (dylandesrosier) [reply test/TwabController.t.sol]: No it won't be exact, these are the average balances and can result in some rounding differences. It should be using the approximate check as well as the `Le`.

File: foundry.toml (hunk 1 above)
Comment (dylandesrosier) [reply foundry.toml]: They're useful to run for invariant tests to get more in depth checks, it's nice to just uncomment rather than look up the keys in case you forget `depth` and `runs`. I don't see any harm in keeping them.

File: src/TwabController.sol (hunk 5 above)
Comment (PierrickGT) [reply src/TwabController.sol:491]: Ok I see. Thanks for the explanation.

File: test/invariants/handlers/TwabControllerHandler.sol (hunk 14 above)
Comment (PierrickGT) [reply test/invariants/handlers/TwabControllerHandler.sol:97]: Ah yes, they only check for address 0.
It's an edge case that would not update the state, so I don't think we need to handle it.

File: test/invariants/handlers/TwabControllerHandler.sol (hunk 16 above)
Comment (PierrickGT) [reply test/invariants/handlers/TwabControllerHandler.sol:278]: I was referring to the `++o_i` instead of `o_i++`

**Review by PierrickGT — APPROVED** [review]

File: src/TwabController.sol (hunk 6 above)
Comment (PierrickGT) [reply src/TwabController.sol:507]: Yes, what I meant is that once you enter this condition, you will necessarily do two writes.
I feel like we duplicate the logic since we have the same kind of logic in `_transferDelegateBalance` but maybe we are ok with it.

Comment (dylandesrosier) [reply src/TwabController.sol:507]: `_transferDelegateBalance` is basically subset of `_transferBalance`. For something like delegating, we can use the more optimized `_transferDelegateBalance` to remove some unnecessary checks.

Doing two writes is unavoidable if the user is delegating to someone other than themselves. But having `_transferBalance` manage updating both the `balance` & `delegateBalance` then we occasionally (likely most of the time since most users will be delegating to themselves) get to do a single write.
